encoding/series_iterator.go

```diff
@@ -0,0 +1,225 @@
+	it := &seriesIterator{pool: pool}
+	it.Reset(id, startInclusive, endExclusive, replicas)
+	return it
+}
+
+func (it *seriesIterator) ID() string {
+	return it.id
+}
+
+func (it *seriesIterator) Start() time.Time {
+	return it.start
+}
+
+func (it *seriesIterator) End() time.Time {
+	return it.end
+}
+
+func (it *seriesIterator) Next() bool {
+	if it.firstNext {
+		it.firstNext = false
+		return it.hasNext()
+	}
+	if !it.hasNext() {
+		return false
+	}
+	it.moveToNext(true)
+	return it.hasNext()
+}
+
+func (it *seriesIterator) Current() (m3db.Datapoint, xtime.Unit, m3db.Annotation) {
+	return it.currIters[0].Current()
+}
+
+func (it *seriesIterator) Err() error {
+	return it.err
+}
+
+func (it *seriesIterator) Close() {
+	if it.isClosed() {
+		return
+	}
+	it.closed = true
+	for _, iter := range it.iters {
+		iter.Close()
+	}
+	it.iters = it.iters[:0]
+	if it.pool != nil {
+		it.pool.Put(it)
+	}
+}
+
+func (it *seriesIterator) Reset(id string, startInclusive, endExclusive time.Time, replicas []m3db.Iterator) {
+	it.id = id
+	it.start = startInclusive
+	it.end = endExclusive
+	it.iters = it.iters[:0]
+	it.currIters = it.currIters[:0]
+	it.unclosedIters = it.unclosedIters[:0]
+	it.err = nil
+	it.firstNext = true
+	it.closed = false
+	for _, replica := range replicas {
+		// Replica can be nil if replica failed to respond
+		if replica == nil {
+			continue
+		}
+		it.iters = append(it.iters, replica)
+	}
+	heap.Init(&it.currIters)
+	it.moveToNext(false)
+}
+
+func (it *seriesIterator) hasError() bool {
+	return it.err != nil
+}
+
+func (it *seriesIterator) isClosed() bool {
+	return it.closed
+}
+
+func (it *seriesIterator) hasMore() bool {
+	return len(it.iters) > 0
+}
+
+func (it *seriesIterator) hasNext() bool {
+	return !it.hasError() && !it.isClosed() && it.hasMore()
+}
+
+func (it *seriesIterator) moveIterToValidNext(iter m3db.Iterator) bool {
+	dp, _, _ := iter.Current()
+	prevT := dp.Timestamp
+	for iter.Next() {
+		dp, _, _ := iter.Current()
+		t := dp.Timestamp
+		if t.Before(prevT) {
+			// Out of order datapoint
+			it.err = errOutOfOrderIterator
+			iter.Close()
+			return false
+		}
+		if t.Before(it.start) {
+			// Fast forward
+			continue
+		}
+		if !t.Before(it.end) {
+			// Close this iter
+			break
+		}
+		return true
+	}
+
+	err := iter.Err()
+	iter.Close()
+	if err != nil {
+		it.err = err
+	}
+	return false
+}
+
+func (it *seriesIterator) moveToNext(pop bool) {
+	if pop {
+		heap.Pop(&it.currIters)
+	}
+
+	for i, iter := range it.iters {
+		for _, existing := range it.currIters {
+			if existing == iter {
+				// Already have this iterator in the currIters
+				continue
+			}
+		}
+
+		if !it.moveIterToValidNext(iter) {
+			// Closed so remove it, two step, first nil out then delete
+			it.iters[i] = nil
+			continue
+		}
+
+		// Dedupe
+		dp, _, _ := iter.Current()
+		t := dp.Timestamp
+		dupe := false
+		for i := range it.currIters {
+			dp, _, _ := it.currIters[i].Current()
+			existsT := dp.Timestamp
```

Comment:
We probably don't want to actually support multiple values for the same timestamp otherwise de-duping them across replica's will become a nightmare.  We should either support a first in or last in wins approach and then let repairs fix any issues arise should multiple concurrent writes across replicas cause inconsistencies for the value for a given timestamp.